### ISSUE TITLE
connection_lost callback?

### ISSUE THREAD

**Issue by danieljfarrell** [Thriftpy/thriftpy2#64]
Transports in [Twisted](https://twistedmatrix.com/documents/current/api/twisted.internet.protocol.Protocol.html#connectionLost) and [asyncio](https://docs.python.org/3.6/library/asyncio-protocol.html#connection-callbacks) have a `connection_lost` callback. Is there an equivalent callback in thiftpy2 which gets called when a transport loses a connection?

**Comment by ethe** [comment]
There is still not yet, I will concern about implement it.

**Comment by danieljfarrell** [comment]
Thank you!

**Comment by ethe** [comment]
I think maybe there are some ways to walk around in current, thriftpy2 will raise a socket exception when connection lost.

**Comment by ethe** [comment]
See https://github.com/Thriftpy/thriftpy2/blob/master/thriftpy2/contrib/aio/socket.py#L342 .

**Comment by danieljfarrell** [comment]
So, for now, one needs to catch a potential socket exception on every call. I suppose that is good practice! 

What do you think about the following?

I could subclass,

https://github.com/Thriftpy/thriftpy2/blob/4065b80530107744172889ca909ce9270ac2f6d6/thriftpy2/contrib/aio/protocol/binary.py#L232

and then implement the [lost_connection](https://docs.python.org/3.6/library/asyncio-protocol.html#asyncio.BaseProtocol.connection_lost) method from the asyncio.BaseProtocol and supply the custom protocol when that when constructing my client 

https://github.com/Thriftpy/thriftpy2/blob/4065b80530107744172889ca909ce9270ac2f6d6/thriftpy2/contrib/aio/rpc.py#L13

**Comment by ethe** [comment]
I think making TAsyncBinaryProtocol inherit from asyncio.BaseProtocol directly is fine to me.

**Comment by danieljfarrell** [comment]
OK, maybe it causes a bit too many changes to inherit from `asyncio.Protocol`. Strictly speaking, it is not required to implement the functionality.

How would you recommend to implement this notification? I'm not roo familiar with socket programming.

For example, could something simple be added here? Such as a callback?

https://github.com/Thriftpy/thriftpy2/blob/f4aa812fd97d1786c51de0c04326dadf1e7dcf16/thriftpy2/contrib/aio/socket.py#L342

**Comment by ethe** [comment]
Yes, it could be, passing into a call back function by the user and check if there is a call back function exists is feasible.
